d2/src/main/java/com/linkedin/d2/balancer/subsetting/DeterministicSubsettingStrategy.java

```diff
@@ -0,0 +1,306 @@
+  private Map<T, Double> _currentWeightedSubset;
+
+  /**
+   * Builds deterministic subsetting strategy
+   *
+   * @param metadataProvider  Provides peer cluster metadata information
+   * @param clusterName The name of the peer cluster
+   * @param minSubsetSize The minimum subset size to satisfy
+   */
+  public DeterministicSubsettingStrategy(DeterministicSubsettingMetadataProvider metadataProvider,
+                                         String clusterName,
+                                         int minSubsetSize,
+                                         LoadBalancerState state)
+  {
+    _metadataProvider = metadataProvider;
+    MD5Hash hashFunction = new MD5Hash();
+    String[] keyTokens = {clusterName};
+    _randomSeed = hashFunction.hashLong(keyTokens);
+    _minSubsetSize = minSubsetSize;
+    _state = state;
+  }
+
+  @Override
+  public boolean isSubsetChanged(long version)
+  {
+    DeterministicSubsettingMetadata metadata = _metadataProvider.getSubsettingMetadata(_state);
+    return metadata == null || !metadata.equals(_currentMetadata) || version != _currentVersion;
+  }
+
+  @Override
+  public Map<T, Double> getWeightedSubset(Map<T, Double> weightMap, long version)
+  {
+    DeterministicSubsettingMetadata metadata = _metadataProvider.getSubsettingMetadata(_state);
+    if (metadata != null)
+    {
+      synchronized (_lock)
+      {
+        if (metadata.equals(_currentMetadata) && version == _currentVersion)
+        {
+          return _currentWeightedSubset;
+        }
+
+        _currentVersion = version;
+        _currentMetadata = metadata;
+
+        List<T> points = new ArrayList<>(weightMap.keySet());
+        Collections.sort(points);
+        Collections.shuffle(points, new Random(_randomSeed));
+        List<Double> weights = points.stream().map(weightMap::get).collect(Collectors.toList());
+        Ring ring = new Ring(weights);
+
+        double offset = metadata.getInstanceId() / (double) metadata.getTotalInstanceCount();
+        double subsetSliceWidth = getSubsetSliceWidth(metadata.getTotalInstanceCount(), points.size());
+        List<Integer> indices = ring.getIndices(offset, subsetSliceWidth);
+
+        _currentWeightedSubset = indices.stream().collect(
+            Collectors.toMap(points::get, i -> round(ring.getWeight(i, offset, subsetSliceWidth), WEIGHT_DECIMAL_PLACE)));
+      }
+      return _currentWeightedSubset;
+    }
+    else
+    {
+      _log.warn("Cannot retrieve metadata required for D2 subsetting. Revert to use all available hosts.");
+      return null;
+    }
+  }
+
+  private static double round(double value, int places)
+  {
+    BigDecimal bd = new BigDecimal(Double.toString(value));
+    bd = bd.setScale(places, RoundingMode.HALF_UP);
+    return bd.doubleValue();
+  }
+
+  private static boolean isEqual(double a, double b, double delta)
+  {
+    return Math.abs(a - b) <= delta;
+  }
+
+  private double getSubsetSliceWidth(int totalClientCount, int totalHostCount)
```

Comment:
To get the subset slice width, should we use server host total weight instead of total count?

Reply:
It should be total count. As mentioned above, the subset slice width is calculated given all hosts have equal weights. It should not change when the host weights are changed. The subset slice width is the same for situation A {host1 = 1.0, host2 = 1.0, host3 = 1.0} and situation B {host1 = 1.0, host2 = 0.0, host3 = 0.0}.

Reply:
Can you use this example to help me understand how is it calculated:
Server side host: 1 redlined host weight: 10.0, 10 other regular host weights: 1.0, so we have total weights as 20.0
MinSubsetSize: 10
Client side: 4 hosts

In this case, what's the subset slice width? I though it should be 1/2? We can have 4 hosts each cover 1/2 of the ring?

Reply:
In your case it would be the entire ring, i.e. no subsetting effect.

Two things to keep in mind: 1. the slice width is calculated given all hosts have weights 1.0.    2. The slice width is a multiple of client unit width = (1 / client count total)

So in your case, minSubsetSize of 10 corresponds to a slice width of 10/11, since there are 11 hosts in total. But since 10/11 is not a multiple of client unit width = 1/4, we will take the next multiple, which is 1.0.

If we have each client covers only 1/2 of the ring, it does not satisfy the minSubsetSize requirement 10/11 when all the hosts return to normal.

Here's a cool visualization of the algorithm https://nvartolomei.com/weighted-deterministic-aperture/. You can play around with weights by clicking the Verbose State drop menu.

Reply:
What confuses me is that what does minSubsettingSize mean, does it mean the actual host number or the weight total (the same question I asked in the pdl change).
If it always means the actual host number, then this is right.

Reply:
It means the actual host number, but in the form of slice width. For a cluster with 100 hosts, minSubsetSize of 10 means 1/10 slice width. It doesn't change when host weight changes.